### PR TITLE
Add /status endpoint

### PR DESCRIPTION
**Pull request by chmeliik**
* CLOUDBLD-2011

Shows status of Cachito workers and services that Cachito depends on.
Also add /status/short, which returns 200 if everything is OK, 503
otherwise.

Signed-off-by: Adam Cmiel <acmiel@redhat.com>

### CODE REVIEW

**Comment by chmeliik** [comment]
Don't let the 550 lines discourage you, 300 of those are unit tests :-)

Note that this will always expose the hostnames (or the RabbitMQ equivalent thereof) of all the Cachito workers, and potentially of other services as well when there is a problem. I don't think that's an issue, but if you can think of a reason why it would be an issue, I can make the output less detailed.

File: cachito/web/status.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    """
+    nexus_status_url = f"{nexus_url.rstrip('/')}/service/rest/v1/status/writable"
+    return _service_ok(nexus_status_url)
+
+
+def athens_ok(athens_url):
+    """
+    Check if Athens is OK.
+
+    Simply checks if the root url is reachable (Athens has no status endpoint).
```
Comment (athos-ribeiro) [inline cachito/web/status.py:50]: Note for future selves: This may be useful in the future https://github.com/gomods/athens/pull/1658

**Comment by chmeliik** [comment]
Found an issue where the worker inspection takes a long time (and then fails) if you kill RabbitMQ, will add separate check for RabbitMQ

File: cachito/web/status.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    for worker_name, reply in reply_tuples:
+        # The reply format does not seem to be documented anywhere, but is
+        # likely a dict with one key: "ok"/"error" and a corresponding message.
+        worker = {"name": worker_name, "ok": "ok" in reply}
+
+        if "ok" not in reply:
+            reason = reply.get("error", "unknown reason")
+            if short:
+                raise CachitoError(f"worker {worker_name} unavailable: {reason}")
```
Comment (athos-ribeiro) [inline cachito/web/status.py:115]: So, if 1 out of 12 workers is unavailable here, we'd respond with an error, but if 11 workers do not respond to the ping above and one do, we'd say everything is ok. Should we onlyreturn 503 if cachito is actually unable to process requests at the moment?

Would adding more information in a 200 response contents help here? Something like "all_systems_go: true/false"?
This of course would only be really truthful if we also included some sort of configuration to let Cachito know the number of workers expected to exist in a certain deployment (this last part may require further discussion).

Comment (athos-ribeiro) [reply cachito/web/status.py:115]: In this matter, should we also consider that Cachito may be partially unavailable? For instance, if Athens is unavailable, Cachito is still operational for python and npm requests.

**Review by athos-ribeiro — COMMENTED** [review]
Looks great!

I added some comments to start some discussions on a couple of topics regarding the new endpoint. I'd be OK with merging this as is and improving the endpoint as the questions and expectations become clearer though.

**Comment by athos-ribeiro** [comment]
> Don't let the 550 lines discourage you, 300 of those are unit tests :-)
> 
> Note that this will always expose the hostnames (or the RabbitMQ equivalent thereof) of all the Cachito workers, and potentially of other services as well when there is a problem. I don't think that's an issue, but if you can think of a reason why it would be an issue, I can make the output less detailed.

Could some of these URLs, e.g., the broker url, include credentials, such as `USERNAME:PASSWORD@HOSTNAME` ?

**Comment by chmeliik** [comment]
> Could some of these URLs, e.g., the broker url, include credentials, such as `USERNAME:PASSWORD@HOSTNAME` ?

Ahh, that is possible, e.g.
```json
{
  "name": "nexus", 
  "ok": false, 
  "reason": "401 Client Error: Unauthorized for url: http://user:pass@nexus:8081/service/rest/v1/status/writable"
}
```

But the only url which contains credentials in our config is the broker one, for that one the error messages do not seem to expose credentials:
```json
{
  "name": "rabbitmq", 
  "ok": false, 
  "reason": "failed to resolve broker hostname"
}
```
```json
{
  "name": "rabbitmq", 
  "ok": false, 
  "reason": "[Errno 111] Connection refused"
}
```
Using invalid credentials in the broker url results in an uncaught exception:
```
amqp.exceptions.AccessRefused: (0, 0): (403) ACCESS_REFUSED - Login was refused using authentication mechanism AMQPLAIN. For details see the broker logfile.
```


But I cannot guarantee that credentials will never be exposed. I see a few options here:
- do not include any error messages, just `"ok": true/false`
- include custom error messages
- censor urls in error messages

**Comment by athos-ribeiro** [comment]
If this is really unknown, I'd rather include custom error messages when needed or censoring the authentication bits of any URLs if they are present. WDYT?

**Comment by chmeliik** [comment]
> If this is really unknown, I'd rather include custom error messages when needed or censoring the authentication bits of any URLs if they are present. WDYT?

Yeah I'm thinking the same thing, currently going with custom, less informative messages

**Comment by chmeliik** [comment]
Updated error messages, using custom format for all errors (better safe then sorry, right?)

File: cachito/web/status.py (hunk 2 above)
Comment (chmeliik) [reply cachito/web/status.py:121]: Hmm, all good points. I think I'll replace the `all_ok` attribute with something like `can_serve`, which will be True if all services and at least one worker are ok.

If we decide to add the configuration for expected number of workers, I'll also add `all_ok`.

I like the idea of partial availability, but I imagine that would best be left for another PR

**Comment by chmeliik** [comment]
Found another issue where the `ping()` method would sometimes inexplicably raise a `ConnectionResetError`, wrapped it in a try/except

File: cachito/web/api_v1.py [hunk 3]
```diff
@@ -22,13 +22,36 @@
...
+        status(short=True)
+        content = ""
+        status_code = 200
+    except CachitoError as e:
+        content = str(e)
+        status_code = 503
+
+    response = make_response(content, status_code)
+    response.content_type = "text/plain"
+    return response
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachito/web/api_v1.py]: OK, done

File: cachito/web/status.py (hunk 2 above)
Comment (chmeliik) [reply cachito/web/status.py]: Done, including partial availability (only for /status, not /status/short)

**Comment by chmeliik** [comment]
Added a few thing:
- custom retries on `ConnectionError` when pinging workers (sometimes, the `ping()` function raises `ConnectionResetError` even when everything is working fine)
- per-package-manager availability info in `/status` response
- only fail if no workers are available, rather than if any workers are unavailable

**Comment by chmeliik** [comment]
Minor unit test change (merged unnecessary tests into one)

File: tests/test_status.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        expect_reason = "503: Service Unavailable"
+    else:
+        expect_reason = None
+
+    ok, reason = status._service_ok(url)
+
+    if fail_reason:
+        assert not ok
+        assert reason == expect_reason
+        assert url not in reason
```
Comment (athos-ribeiro) [inline tests/test_status.py:44]: :rocket:

File: tests/test_status.py [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+            {"gomod": False, "npm": False, "pip": False, "git-submodule": True},
+        ),
+    ],
+)
+def test_can_process(worker_ok, failing_services, expect_result):
+    services = [{"name": service, "ok": False} for service in failing_services]
+    assert status._can_process(TEST_PACKAGE_MANAGERS, services, worker_ok) == expect_result
+
+
+def mock_worker_config(nexus_hoster=False):
```
Comment (athos-ribeiro) [inline tests/test_status.py:217]: Nitpick: could we move this function to the beginning of the file?

**Review by athos-ribeiro — APPROVED** [review]
Great work!!!

Added one comment for moving a mocking function to the beginning of a module. This should not be blocker. Feel free to merge with or without addressing the comment :)

File: tests/test_status.py (hunk 5 above)
Comment (chmeliik) [reply tests/test_status.py]: Done

**Pull request merged**